packages/runtime-core/src/components/KeepAlive.ts

```diff
@@ -228,7 +228,16 @@ const KeepAliveImpl: ComponentOptions = {
     const cacheSubtree = () => {
       // fix #1621, the pendingCacheKey could be 0
       if (pendingCacheKey != null) {
-        cache.set(pendingCacheKey, getInnerChild(instance.subTree))
+        // if KeepAlive child is a Suspense, it needs to be cached after Suspense resolves
+        // avoid caching vnode that not been mounted
+        if (isSuspense(instance.subTree.type)) {
+          const cacheKey = pendingCacheKey
+          queuePostRenderEffect(() => {
+            cache.set(cacheKey, getInnerChild(instance.subTree))
```

Comment:
```suggestion
          queuePostRenderEffect(() => {
            cache.set(pendingCacheKey, getInnerChild(instance.subTree))
```

Reply:
Could the `pendingCacheKey` change after the `Suspense` is resolved?
In that case, @edison1105's implementation is the correct one. There isn't a test case covering this edge case, though.

Reply:
> Could the pendingCacheKey change after the Suspense is resolved?

This scenario won't occur. If the KeepAlive child switches (pendingCacheKey changes) before the Suspense resolves, the Suspense won't resolve. I wrote it that way for safety considerations, but it seems I overthought it, and it's unnecessary.